### ISSUE TITLE
Sighup reconfiguration can disable gitpoller database connection

### ISSUE THREAD

**Issue by YngveNPettersen** [buildbot/buildbot#4673]
We have had several incidents where sighup reconfiguration of the master have caused it to stop handling new commits; IIRC the reconfigurations have involved new branches and schedulers.

The pre-condition for the problem seems to be that the polling started or was ongoing at the time the reconfiguration operation was performed.

The problem does not seem to affect ongoing or scheduled jobs on the builders and workers, but to get the gitpolling working again a stop/restart operation have to be performed.

~~There are also an possibility that this problem can cause handling of commits from a branch to become completely broken. Updates to the branch is  registered in git, the scheduler exists, but it does not look like the the changesource is handling the new commits.~~ (Moved to https://github.com/buildbot/buildbot/issues/4686 )

~~After several attempts involving stop/restart of the master, creation of new branches and schedulers without any luck (although this may have been due to the new changesources not being ready, log information about state of a branch changesource  would be very useful so that we could know *when* the changesource is ready for use; this information would also be even more useful if it was available in the GUI). After these failed attempts I reverted my changes and returned to the old configuration for this branch and its two schedulers. The next day I restarted the master for other reasons, and suddenly the schedulers were finally working as they should.~~

I have observered a database related exception during these incidents, this is the most recent one, which was likely involved in the above scheduler problem:

```
2019-03-16 13:02:31+0000 [-] while polling for changes
  Traceback (most recent call last):
    File "sandbox/lib/python3.6/site-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
      result = g.send(result)
    File "sandbox/lib/python3.6/site-packages/buildbot/changes/gitpoller.py", line 233, in poll
      yield self.setState('lastRev', self.lastRev)
    File "sandbox/lib/python3.6/site-packages/twisted/internet/defer.py", line 1613, in unwindGenerator
      return _cancellableInlineCallbacks(gen)
    File "sandbox/lib/python3.6/site-packages/twisted/internet/defer.py", line 1529, in _cancellableInlineCallbacks
      _inlineCallbacks(None, g, status)
  --- <exception caught here> ---
    File "sandbox/lib/python3.6/site-packages/buildbot/changes/gitpoller.py", line 233, in poll
      yield self.setState('lastRev', self.lastRev)
    File "sandbox/lib/python3.6/site-packages/twisted/internet/defer.py",
line 1418, in _inlineCallbacks
      result = g.send(result)
    File "sandbox/lib/python3.6/site-packages/buildbot/util/state.py", line 43, in setState
      yield self.master.db.state.setState(self._objectid, key, value)
  builtins.AttributeError: 'NoneType' object has no attribute 'db'
 
2019-03-16 13:02:31+0000 [-] Caught exception while deactivating ClusteredService(...)
  Traceback (most recent call last):
    File "sandbox/lib/python3.6/site-packages/twisted/internet/defer.py", line 654, in _runCallbacks
      current.result = callback(current.result, *args, **kw)
    File "sandbox/lib/python3.6/site-packages/twisted/internet/defer.py", line 1475, in gotResult
      _inlineCallbacks(r, g, status)
    File "sandbox/lib/python3.6/site-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
      result = g.send(result)
    File "sandbox/lib/python3.6/site-packages/buildbot/util/service.py", line 341, in stopService
      log.err(e, _why="Caught exception while deactivating ClusteredService(%s)" % self.name)
  --- <exception caught here> ---
    File "sandbox/lib/python3.6/site-packages/buildbot/util/service.py", line 339, in stopService
      yield self._unclaimService()
    File "sandbox/lib/python3.6/site-packages/buildbot/changes/base.py", line 51, in _unclaimService
      return self.master.data.updates.trySetChangeSourceMaster(self.serviceid,
  builtins.AttributeError: 'NoneType' object has no attribute 'data'
```

**Comment by YngveNPettersen** [comment]
A bit more about the schedulers and associated branches:

* There are two branches, both tracking the same branch. In one phase one of them is the leading one, later the other, but they may point at the same commit. We expect builds to be generated for both branches, even when they point at the same commit.

* Each branch is associated with two schedulers, one that triggers "immediately" when a new commit is pushed, the other is a Nightly scheduler. The schedulers all have different names.

At present it seems as if the change source is the same for both branches for each type of scheduler.

That is, if I push a commit to Branch A, the schedulers for that branch will trigger. If I move Branch B to the same commit, and push it, its schedulers will not trigger. Similarly, if I push a new commit to Branch B, its schedulers will trigger, while Branch A's schedulers will not trigger when I move it to Branch B's commit and push it.

It seems the schedulers are fetching commits from the same list of commits, despite them being for different branches.

This is the first time we are actually deploying this configuration fully on the new builder (we were only moving one branch after moving to this new buildbot version), the old 0.8.4/chromium-based buildbot had no problems with this configuration.

Either v2.0.1 will not trigger a new build on a commit for a specific class of scheduler that it has already triggered a build for, or the database hickup mentioned above caused the the configuration to be incorrect. Neither is desirable to us.

Assuming it is the first case (which is what it is starting to look like), our expectation is that there will be a build for each unique combination (repository,  schedulername, branchname, commitID), with several commits from the same (repository, schedulename, branch) being collected into a single build being acceptable.

We also have a group of other schedulers that work on the same branch, including two groups of Nightlies, for different builders. Although I have not observed this situation, it is not desirable that a build triggered by one nightly scheduler might prevent a build for the same commit by another scheduler.

**Comment by tardyp** [comment]
Hi @YngveNPettersen we need need to be a bit more rigourous in the issues, and create on item by issue. Here it looks like there are several problems.
Please be specific and concise or you are at risk of people TL;DR you. Remember that Buildbot is a voluntary project.

**Comment by YngveNPettersen** [comment]
The reason I, so far, have had both issues in this report is that I originally connected the scheduler problem to the database hickup issue and the hickup being the cause of the scheduler issue. While I am now leaning in the direction of it being a separate issue, I am not yet fully convinced the database hickup didn't cause the scheduler issue. It will, however, be a month or two until I add a similar production configuration.

Can you confirm that the currently implemented behavior is that unique (repository, scheduler_kind, commitID) get triggered for a scheduler, not (repository, scheduler_name, branch, commitID)?

**Comment by tardyp** [comment]
I can't confirm that, this depends on how you configure your schedulers, and change filters

**Comment by YngveNPettersen** [comment]
The current configuration is 

* a GitPoller for the repo URL, and the list of branches
* Each scheduler is initialized with the name of the scheduler, a list of builders, and a ChangeFilter for the branch. The schedulers in this case is either a SingleBranchScheduler or a schedulers.Nightly

**Comment by pitmanst** [comment]
I see the same issue as mentioned in the description - doing a reconfig during polling causes it to disconnect on that given repository (other repositories will still poll fine). Here's a way that you can reproduce the error using the default buildbot config that's provided:

```
#!/bin/bash

# Cleanup if it needs to run multiple times
rm -rf ./sandbox ./master

# Installation
python3 -m venv sandbox
source ./sandbox/bin/activate
pip3 install 'buildbot[bundle]'

# Create default master
buildbot create-master master
mv master/master.cfg.sample master/master.cfg
sed -i.sedtmp 's#pollInterval=300#pollInterval=3#g' master/master.cfg # Not needed, just speeds up reproducing
buildbot start master

# Reconfig at the same time as master polls
for i in {1..10}
do
  sleep 6
  buildbot reconfig master
done

# Give buildbot some time to poll and log it in twistd.log if it's working
sleep 12
buildbot stop master
```

It's unfortunately not a guaranteed way to reproduce and you may need to run it a few times. But I can see it reproduce on both my Mac and a Linux machine. If you see the error, you can check twistd.log and notice that's no longer polling.

**Comment by ciarancourtney** [comment]
+1 seeing this more and more often

**Comment by ciarancourtney** [comment]
FYI for the moment this works for me, at the end of a build I touch a file in `/var/tmp`, a cron then runs the below every minute if the file exists and deletes it on success.

    curl --silent --show-error <master url>/buildbot/api/v2/builds?complete=false | grep -q 'builds": \[\]' && sudo systemctl restart buildbot

Buildbot is restared if the JSON response key `builds` is an empty array.

**Comment by YngveNPettersen** [comment]
FYI, we had this happen again just last week, when we did a scheduler update

**Comment by hmoffatt** [comment]
I have seen this a few times, including just recently. I have 2.10.1.

```
2021-02-19 15:13:58+1100 [-] gitpoller: processing changes from "file:///srv/svn/webapp-ui.git"
2021-02-19 15:14:13+1100 [-] gitpoller: processing changes from "file:///srv/svn/webapp-ui.git"
2021-02-19 15:14:14+1100 [-] beginning configuration update
2021-02-19 15:14:14+1100 [-] Loading configuration from '/srv/buildbot/master/master.cfg'
2021-02-19 15:14:14+1100 [-] while polling for changes
        Traceback (most recent call last):
          File "/srv/buildbot/sandbox/lib/python3.7/site-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
            result = g.send(result)
          File "/srv/buildbot/sandbox/lib/python3.7/site-packages/buildbot/changes/gitpoller.py", line 226, in poll
            yield self.setState('lastRev', self.lastRev)
          File "/srv/buildbot/sandbox/lib/python3.7/site-packages/twisted/internet/defer.py", line 1613, in unwindGenerator
            return _cancellableInlineCallbacks(gen)
          File "/srv/buildbot/sandbox/lib/python3.7/site-packages/twisted/internet/defer.py", line 1529, in _cancellableInlineCallbacks
            _inlineCallbacks(None, g, status)
        --- <exception caught here> ---
          File "/srv/buildbot/sandbox/lib/python3.7/site-packages/buildbot/changes/gitpoller.py", line 226, in poll
            yield self.setState('lastRev', self.lastRev)
          File "/srv/buildbot/sandbox/lib/python3.7/site-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
            result = g.send(result)
          File "/srv/buildbot/sandbox/lib/python3.7/site-packages/buildbot/util/state.py", line 43, in setState
            yield self.master.db.state.setState(self._objectid, key, value)
        builtins.AttributeError: 'NoneType' object has no attribute 'db'
        
2021-02-19 15:14:14+1100 [-] Caught exception while deactivating ClusteredService(file:///srv/svn/webapp-ui.git)
        Traceback (most recent call last):
          File "/srv/buildbot/sandbox/lib/python3.7/site-packages/twisted/internet/defer.py", line 654, in _runCallbacks
            current.result = callback(current.result, *args, **kw)
          File "/srv/buildbot/sandbox/lib/python3.7/site-packages/twisted/internet/defer.py", line 1475, in gotResult
            _inlineCallbacks(r, g, status)
          File "/srv/buildbot/sandbox/lib/python3.7/site-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
            result = g.send(result)
          File "/srv/buildbot/sandbox/lib/python3.7/site-packages/buildbot/util/service.py", line 347, in stopService
            log.err(e, _why=msg)
        --- <exception caught here> ---
          File "/srv/buildbot/sandbox/lib/python3.7/site-packages/buildbot/util/service.py", line 344, in stopService
            yield self._unclaimService()
          File "/srv/buildbot/sandbox/lib/python3.7/site-packages/buildbot/changes/base.py", line 51, in _unclaimService
            return self.master.data.updates.trySetChangeSourceMaster(self.serviceid,
        builtins.AttributeError: 'NoneType' object has no attribute 'data'
        
```

**Comment by p12tic** [comment]
@hmoffatt I think the exceptions you are seeing may have been fixed in 0328270495772a29b50763ef9f3e90043fad441a. Gitpoller is likely to have other issues that break it during reconfiguration, I plan to look into this in near future.

**Comment by p12tic** [comment]
I believe this bug has been fixed in two separate bug fixes: #5795 and #5798 which have been released as 2.10.2 and 3.0.0. Thus I'm closing the issue. Please reopen if the bug is still exists. Thanks a lot!
